### ISSUE TITLE
Extending from main application

### ISSUE THREAD

**Issue by loevgaard** [hostnet/entity-plugin-lib#52]
Hi

I have [this package](https://github.com/loevgaard/dandomain-foundation-entities), which has the entities I use in my main application.

Is there any way I can create a `OrderTrait` in my main application that gets included in the `Loevgaard\DandomainFoundation\Entity\Generated\OrderTrait`? In your readme you mention the `entity-bundle-dir`, but there is not example, and I can't seem to get it working. I tried to set this value to `src/AppBundle/Entity`, but then the plugin generated all the traits and interfaces with that folder.

**Comment by nicoschoenmaker** [comment]
Example from one of our own apps `composer.json`
```
{
    "extra": {
        "entity-bundle-dir": "src/Hostnet/Bundle/MyBundle/"
    }
}
```

So without `Entity`, in the `extra` section :slightly_smiling_face:

**Comment by nicoschoenmaker** [comment]
The reason it hints the bundle dir instead of the entity dir is because it has the same system for extending repositories as well.

**Comment by loevgaard** [comment]
so if I want to extend my `Order` in the entities package I create a `OrderTrait` within `src/Hostnet/Bundle/MyBundle/`?

**Comment by loevgaard** [comment]
Does the namespace has to match? I.e. should the namespace of `OrderTrait` match the namespace of `Order`?

**Comment by nicoschoenmaker** [comment]
```
"entity-bundle-dir": "src/Hostnet/Bundle/MyBundle/"
```
```
// src/Hostnet/Bundle/MyBundle/Entity/OrderTrait.php
namespace Hostnet\Bundle\MyBundle\Entity;
```

**Comment by loevgaard** [comment]
Okay. Sorry for being a pain in the ass. I have the following now:

```js
// composer.json for main application
{
    // ...
    "type": "project",
    "autoload": {
        "psr-4": {
            "": "src/",
            "Loevgaard\\DandomainFoundation\\": "src/Loevgaard/DandomainFoundation"
        },
        "classmap": [ "app/AppKernel.php", "app/AppCache.php" ]
    },
    // ...
    "require": {
        // ...
        // this is the library that requires the entities package
        "loevgaard/dandomain-foundation-bundle": "dev-implement-hostnet-entities",
        // ...
    },
    // ...
    "extra": {
        "symfony-app-dir": "app",
        "symfony-bin-dir": "bin",
        "symfony-var-dir": "var",
        "symfony-web-dir": "web",
        "symfony-tests-dir": "tests",
        "symfony-assets-install": "relative",
        "incenteev-parameters": {
            "file": "app/config/parameters.yml"
        },
        "entity-bundle-dir": "src/Loevgaard/DandomainFoundation/"
    }
}
```

```php
// src/Loevgaard/DandomainFoundation/Entity/OrderLineTrait.php
<?php

namespace Loevgaard\DandomainFoundation\Entity;

class OrderLineTrait
{
    public function getTest() {
        return 'test';
    }
}
```

When I run `composer dump-autoload` it doesn't include this one. It is [this entity](https://github.com/loevgaard/dandomain-foundation-entities/blob/master/src/Entity/OrderLine.php) I am trying to extend.

**Comment by nicoschoenmaker** [comment]
I just created a project with two files. 1. a `composer.json` like
```
{
    "autoload": {
        "psr-4": {
            "Loevgaard\\DandomainFoundation\\": "src/Loevgaard/DandomainFoundation"
        }
    },
    "require": {
        "loevgaard/dandomain-foundation-bundle": "dev-implement-hostnet-entities",
        "loevgaard/dandomain-foundation-entities": "dev-master"
    },
    "extra": {
        "entity-bundle-dir": "src/Loevgaard/DandomainFoundation/"
    }
}
```

2. My entity trait at `src/Loevgaard/DandomainFoundation/Entity/OrderLineTrait.php` like
```
<?php

namespace Loevgaard\DandomainFoundation\Entity;

trait OrderLineTrait
{
    public function getTest()
    {
        return 'test';
    }
}
```

Then I get a `./vendor/loevgaard/dandomain-foundation-entities/src/Entity/Generated/OrderLineTrait.php` like
```
<?php
namespace Loevgaard\DandomainFoundation\Entity\Generated;

use Loevgaard\DandomainFoundation\Entity\OrderLineTrait as LoevgaardDandomainFoundationEntity;

/**
 * Trait of traits generated for OrderLine
 * This is the guy that the main class needs to require, and ensures that everything is glued
 * together
 */
trait OrderLineTrait
{
    use LoevgaardDandomainFoundationEntity;
}
```

Seems to work for me. Not sure what I'm doing different though. I would recommend against using the same name space in a package and in the main app though, that can get confusing real fast.

Good news is that the entity package itself is working, so maybe something in the app?

**Comment by loevgaard** [comment]
Thank you very much Nico! I will try this! Looks like you required the `loevgaard/dandomain-foundation-entities` in the composer.json of your main application. Does this have an effect maybe?

**Comment by loevgaard** [comment]
Hi Nico

Just tested it now. It works only when you require the entities in the main application composer.json. Thank you very much for your help!

**Comment by nicoschoenmaker** [comment]
No problem :) happy it works for you!

The plugin will only glue traits in when the targets are required (or suggested).

This is helpful when you have two entities with the same name (different namespace). In that case you can denote with your require to which one you want to glue your trait.

**Comment by loevgaard** [comment]
Makes sense :-)

**Comment by loevgaard** [comment]
Hi @nicoschoenmaker 

You wrote this:

> The reason it hints the bundle dir instead of the entity dir is because it has the same system for extending repositories as well.

How do you use the plugin to extend repositories? :)

**Comment by nicoschoenmaker** [comment]
Heya,

TL;DR; Same way as an entity.

In the entity package
```php
class MyRepository
{
    use Generated\MyRepositoryTrait;
}
```

In an other package (or app) that depends on the entity package
```php
trait MyRepositoryTrait
{
}
```

Then it will be included.

**Comment by loevgaard** [comment]
Okay, that's what I thought, but I guess you move a lot of logic over to the entity package. But anyways I think it's a good solution.

Do I need a folder named 'Repository' for the repository classes?

**Comment by nicoschoenmaker** [comment]
Yeah, it's based on the Symfony directory structure where you place entities in the `Entity` namespace and repositories in `Repository`. That happens here in the source: https://github.com/hostnet/entity-plugin-lib/blob/master/src/EntityPackageBuilder.php#L49-L56

**Comment by loevgaard** [comment]
Thank you very much, Nico! You have been a great help!

**Comment by loevgaard** [comment]
Hi Nico

Is there a reason that you don't generate interfaces for repositories?

**Comment by nicoschoenmaker** [comment]
Heya,

Short answer: we do not miss them yet.

Longer answer: It basically works without the interfaces. We originally added the interfaces
- so there was one place where you can see what you can do with the entity. This becomes vague when you're using a bunch of traits, some of which might be included conditionally.
- so there would be auto completion (at the time our IDE didn't fully support traits.)

The latter has been fixed now, the first reason didn't motivate us enough to add them. Would be open for a PR if you'd be interested, though.

**Comment by loevgaard** [comment]
It would be nice for testing purposes and also if people wanted to implement the repository interface instead of using mine.

I haven't created composer plugins before, so I don't know where to start in your code. Can you point me in the right direction?

**Comment by nicoschoenmaker** [comment]
Some tips off the top of my head. The plugin is not that big actually :slightly_smiling_face: 

The party starts at the [`Plugin`](https://github.com/hostnet/entity-plugin-lib/blob/master/src/Plugin.php#L19-L41). It basically creates the object tree (container-ish) & puts the `Installer` to work.

The library works based on reflection. When creating a `new ReflectionClass` for a class that implements an interface that does not exist (yet) php crashes. 

That is why plugin does it's work in three phases:
- Phase 1 is already working. It generates `Generated/FooTrait.php` / `Generated/FooRepositoryTrait`.
- Phase 2 generates an [empty](https://github.com/hostnet/entity-plugin-lib/blob/master/src/EmptyGenerator.php) interface. That happens in [`Installer::generateEmptyCode`](https://github.com/hostnet/entity-plugin-lib/blob/master/src/Installer.php#L198-L219).
- Phase 3 fills the interfaces with the actual methods. That happens in [`Installer:generateConcreteIndividualCode`](https://github.com/hostnet/entity-plugin-lib/blob/master/src/Installer.php#L221-L241) and then the [`ReflectionGenerator::generate`](https://github.com/hostnet/entity-plugin-lib/blob/master/src/ReflectionGenerator.php#L30-L53)

What is going wrong right now is that `Installer::generateEmptyCode` and `Installer::generateConcreteIndividualCode` only look at entities, not repositories.

**Comment by yannickl88** [comment]
Small sidenote, I would advise against having interfaces be generated which are for 3rd party developers.

Firstly, because you are depended on the code generations, changes in that could break your *public API* and cause it to be BC break in your package. 
Secondly, it because very fuzzy for 3rd party developers to actually find them in your source code. It requires a working checkout to see the interface.
And finally, you have less control over migrations and BC breaks. A non 1-to-1 interface of your class allows you the flexibility to already prepare for future changes without changing your public API.

**Comment by nicoschoenmaker** [comment]
One unexpected thing we saw when using these generated interfaces:
When you add a method to your repository, it will add a method in the interface. If someone else has implemented this interface, they will get this error
```
PHP Fatal error:  Class CustomRepository contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (MyRepositoryInterface::myMethod)
```
Adding a method becomes a bc break with the interfaces,
